internal/server/middleware.go

```diff
@@ -182,3 +183,89 @@ func getCookie(req *http.Request, name string) (string, error) {
 	}
 	return url.QueryUnescape(cookie.Value)
 }
+
+func OrganizationFromDomain(defaultOrgName, defaultOrgDomain string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		org, host, err := getOrgFromRequest(c, defaultOrgName, defaultOrgDomain)
+		if err != nil {
+			sendAPIError(c, err)
+			return
+		}
+		if org != nil {
+			c.Set("host", host)
+			c.Set("org", org)
+			logging.Debugf("organization set to %s for host %s", org.Name, host)
+		}
+		c.Next()
+	}
+}
+
+// returns org, host
+func getOrgFromRequest(c *gin.Context, defaultOrgName, defaultOrgDomain string) (*models.Organization, string, error) {
+	db := getDB(c)
+	host := c.Request.Header.Get("Host")
+
+	if len(host) == 0 {
+		// tests are lazy and don't set Host
+		host = defaultOrgDomain
+	}
+
+	if len(host) > 0 {
+	hostLookup:
+		org, err := data.GetOrganization(db, data.ByDomain(host))
+		if err != nil {
+			if errors.Is(err, internal.ErrNotFound) {
+				// first, remove port and try again
+				h, p, err := net.SplitHostPort(host)
+				if len(p) > 0 && err == nil {
+					host = h
+					goto hostLookup
+				}
+
+				org, err = getDefaultOrg(db, defaultOrgName, defaultOrgDomain)
+				if err != nil {
+					return nil, "", fmt.Errorf("creating default organization: %w", err)
+				}
+				return org, host, nil
+			}
+			logging.Errorf("fetching org: %s", err)
+			return nil, "", nil
+		}
+		return org, host, nil
+	}
+	return nil, "", nil
+}
+
+func OrganizationRequired() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		o, ok := c.Get("org")
+		if !ok {
+			sendAPIError(c, internal.ErrBadRequest)
+			return
+		}
+		if _, ok = o.(*models.Organization); !ok {
+			sendAPIError(c, internal.ErrBadRequest)
+			return
+		}
+		c.Next()
+	}
+}
+
+var defaultOrgCache *models.Organization
```

Comment:
I hope we can avoid this. #1939 is an issue to track removing exactly this pattern from other places we've done it. These package-level static vars are a real risk for testing. They require `Patch` functions to clear one the cache, and it's easy to miss and hide bugs.

Could we cache this on the `Server` type instead? Maybe the defaults and this cache could be part of some type stored as a field on `Server` ?

That would give us a nice pattern for fixing #1939.

We'll need to address some part of that for org support anyway, because currently those caches are not org-aware.